searchlib/src/tests/queryeval/nearest_neighbor/nearest_neighbor_test.cpp

```diff
@@ -0,0 +1,177 @@
+
+using search::feature_t;
+using search::tensor::DenseTensorAttribute;
+using search::AttributeVector;
+using vespalib::eval::ValueType;
+using vespalib::eval::TensorSpec;
+using vespalib::tensor::Tensor;
+using vespalib::tensor::DenseTensorView;
+using vespalib::tensor::DefaultTensorEngine;
+
+using namespace search::fef;
+using namespace search::queryeval;
+
+vespalib::string denseSpec("tensor(x[2])");
+
+std::unique_ptr<DenseTensorView> createTensor(const TensorSpec &spec) {
+    auto value = DefaultTensorEngine::ref().from_spec(spec);
+    DenseTensorView *tensor = dynamic_cast<DenseTensorView*>(value.get());
+    ASSERT_TRUE(tensor != nullptr);
+    value.release();
+    return std::unique_ptr<DenseTensorView>(tensor);
+}
+
+struct Fixture
+{
+    using BasicType = search::attribute::BasicType;
+    using CollectionType = search::attribute::CollectionType;
+    using Config = search::attribute::Config;
+
+    Config _cfg;
+    vespalib::string _name;
+    vespalib::string _typeSpec;
+    std::shared_ptr<DenseTensorAttribute> _tensorAttr;
+    std::shared_ptr<AttributeVector> _attr;
+
+    Fixture(const vespalib::string &typeSpec)
+        : _cfg(BasicType::TENSOR, CollectionType::SINGLE),
+          _name("test"),
+          _typeSpec(typeSpec),
+          _tensorAttr(),
+          _attr()
+    {
+        _cfg.setTensorType(ValueType::from_spec(typeSpec));
+        _tensorAttr = makeAttr();
+        _attr = _tensorAttr;
+        _attr->addReservedDoc();
+    }
+
+    ~Fixture() {}
+
+    std::shared_ptr<DenseTensorAttribute> makeAttr() {
+        return std::make_shared<DenseTensorAttribute>(_name, _cfg);
+    }
+
+    void ensureSpace(uint32_t docId) {
+        while (_attr->getNumDocs() <= docId) {
+            uint32_t newDocId = 0u;
+            _attr->addDoc(newDocId);
+            _attr->commit();
+        }
+    }
+
+    void setTensor(uint32_t docId, const Tensor &tensor) {
+        ensureSpace(docId);
+        _tensorAttr->setTensor(docId, tensor);
+        _attr->commit();
+    }
+
+    void setTensor(uint32_t docId, double v1, double v2) {
+        auto t = createTensor(TensorSpec(denseSpec)
+                              .add({{"x", 0}}, v1)
+                              .add({{"x", 1}}, v2));
+        setTensor(docId, *t);
+    }
+};
+
+template <bool strict>
+SimpleResult find_matches(Fixture &env) {
+    using NNI = NearestNeighborIterator<strict>;
+    auto md = MatchData::makeTestInstance(2, 2);
+    auto qt = createTensor(TensorSpec(denseSpec));
+
+    auto &tfmd = *(md->resolveTermField(0));
+    const DenseTensorView &qtv = *qt;
+    auto &attr = *(env._tensorAttr);
+
+    NearestNeighborDistanceHeap dh(2);
+    NNI search(tfmd, qtv, attr, dh);
+    if (strict) {
+        return SimpleResult().searchStrict(search, attr.getNumDocs());
+    } else {
+        return SimpleResult().search(search, attr.getNumDocs());
+    }
+}
+
+TEST("require that NearestNeighborIterator returns expected results") {
```

Comment:
Consider adding a test case where the query point is not in origo.

Reply:
done.